### ISSUE TITLE
Discrepancy of batch norm to A0 paper

### ISSUE THREAD

**Issue by Cyanogenoid** [LeelaChessZero/lczero-training#28]
In [A0](https://arxiv.org/abs/1712.01815), 64 TPU workers are used to train the network. That means that with their batch size of 4096, each worker independently processes a subset of every minibatch containing 64 samples. Thus, the batch statistics of the batch norms that each worker computes is actually over these 64 samples, not over the whole minibatch. Our current training code computes batch norm statistics over the whole minibatch.

In [this paper](https://arxiv.org/abs/1705.08741), "ghost batch normalisation" is discussed, which is a scheme equivalent to splitting the batch to multiple workers as was done in A0. They find that this improves generalisation performance (see Table 1 "+GBN" column), which should translate to better playing strength for Leela.

My suggestion is to change the training code to include ghost batch normalisation in order to match what is described in the A0 paper. Fortunately, this is simple to implement because [`tf.layers.batch_normalization`](https://www.tensorflow.org/api_docs/python/tf/layers/batch_normalization) supports it through the `virtual_batch_size` argument. In the three places in tfprocess.py where it is called, it simply needs another argument `virtual_batch_size=64` passed to it. No other changes should be needed.

**Comment by GBeauregard** [comment]
As long as we are in agreement about A0 doing this I support making this change for future runs, but I would be against merging this into the test 10 pipeline (by bootstrapping or otherwise). It should be in test 11+.

**Comment by ASilver** [comment]
Is there a way to generate a larger batch by doing minibatches? I.e. the 1080ti cannot handle a batch over 512 in a 20b due to memory limits, but it could do them sequentially and then emulate the A0 process by then 'joining them' into a larger batch of 2048 or even 4096.

**Comment by Cyanogenoid** [comment]
Some more support for A0 using batch norm in the way I described: 

- In the [AlphaGo Zero](https://www.nature.com/articles/nature24270.epdf?author_access_token=VJXbVjaSHxFoctQQ4p2k4tRgN0jAjWel9jnR3ZoTv0PVW4gB86EEpGqTRDtpIz-2rmo8-KG06gqVobU5NSCFeHILHcVFUeMsbvwS-lxjqQGg98faovwjxeTUgZAUMnRQ) paper in the optimization section, they talk about using 64 GPU workers (same as A0) working on 32 size mini-batches each, for a total mini-batch of size 2,048. A0 uses a total mini-batch size of 4,096, so it makes sense that each worker is processing mini-batches of size 64 in this case.
- In the Ghost Batch Norm paper linked above, they say:
> We note that in a multi-device distributed setting, some of the benefits of "Ghost BN" may already
> occur, since batch-normalization is often preformed on each device separately to avoid additional
> communication cost. Thus, each device computes the batch norm statistics using only its samples
> (i.e., part of the whole mini-batch).

**Comment by Cyanogenoid** [comment]
@ASilver Yes, it would involve accumulating gradients several times and only then doing a gradient update with the accumulated gradients. In Tensorflow, it would look something like this: https://stackoverflow.com/questions/46772685/how-to-accumulate-gradients-in-tensorflow
This does seem a bit out-of-scope from this issue though, so creating a separate issue might be better suited for that.
